pkg/querier/worker/scheduler_processor_test.go

```diff
@@ -402,6 +403,47 @@ func TestSchedulerProcessor_ResponseStream(t *testing.T) {
 			require.Equal(t, tc.responseBodyBytes, frontend.responses[queryID].body)
 		})
 	}
+
+	t.Run("should abort streaming if query is cancelled", func(t *testing.T) {
+		sp, loopClient, requestHandler, frontend := prepareSchedulerProcessor(t)
+		sp.streamingEnabled = true
+		sp.maxMessageSize = 5 * responseStreamingBodyChunkSizeBytes
+
+		recvCount := atomic.NewInt64(0)
+		frontend.responseStreamStarted = make(chan struct{})
+
+		queryID := uint64(1)
+		loopClient.On("Recv").Return(func() (*schedulerpb.SchedulerToQuerier, error) {
+			switch recvCount.Inc() {
+			case 1:
+				return &schedulerpb.SchedulerToQuerier{
+					QueryID:         queryID,
+					FrontendAddress: frontend.addr,
+					UserID:          "test",
+				}, nil
+			default:
+				<-frontend.responseStreamStarted
+				loopClient.cancelCtx()
+				return nil, toRPCErr(context.Canceled)
+			}
+		})
+
+		workerCtx, workerCancel := context.WithCancel(context.Background())
+		responseBodySize := 4*responseStreamingBodyChunkSizeBytes + 1
+		requestHandler.On("Handle", mock.Anything, mock.Anything).Run(func(args mock.Arguments) {
+			workerCancel()
```

Comment:
doesn't this test two kinds of cancels - worker cancel (where we mustn't abort the stream) and query cancel (where it's ok to abort the stream)? If yes, then let's test them in two separate cases because now the query cancel overshadows the worker cancel

Reply:
Yes, I have changed the logic to run `processQueriesOnSingleStream` in a goroutine and cancel the worker context only after completion of the tests.